packages/utils/src/lib/server/cache/s3.ts

```diff
@@ -0,0 +1,171 @@
+import { BaseCache } from './base';
+import {
+	DeleteObjectCommand,
+	DeleteObjectsCommand,
+	GetObjectCommand,
+	HeadObjectCommand,
+	ListObjectsV2Command,
+	PutObjectCommand,
+	S3,
+	type S3ClientConfig,
+} from '@aws-sdk/client-s3';
+import { logger } from '../../logger.js';
+import { createJitter, JitterMode, type JitterFn } from '../../utils/misc.js';
+
+type S3CacheOptions = {
+	bucket: string;
+	keyPrefix?: string;
+	defaultTTL?: number;
+	defaultJitter?: JitterMode | JitterFn;
+};
+
+export class S3Cache<V extends Uint8Array = Uint8Array> extends BaseCache<V> {
+	readonly #options: S3CacheOptions;
+	readonly #client: S3;
+
+	public static init<V extends Uint8Array = Uint8Array>(
+		s3Options: S3ClientConfig,
+		options: S3CacheOptions,
+	): S3Cache<V> {
+		const client = new S3(s3Options);
+
+		return new this(options, client);
+	}
+
+	private constructor(options: S3CacheOptions, client: S3) {
+		super();
+
+		this.#options = options;
+		this.#client = client;
+	}
+
+	private buildKey(key: string): string {
+		return `${this.#options.keyPrefix ?? ''}${key}`;
+	}
```

Comment:
Can we use a private method instead?

```suggestion
	#buildKey(key: string): string {
		return `${this.#options.keyPrefix ?? ''}${key}`;
	}
```